### DISCUSSION TITLE
Spec/Discussion: Advanced Corner Radius

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#7923] [General]
CornerRadius is a relatively new addition to Avalonia based on the UWP implementation. I've also just added RadiusX/Y to Rectangle based on the WPF implementation. I've noticed a considerable difference in thinking behind these properties that needs to be discussed.

This current implementations are as follows:

 1. `CornerRadius` : This implementation allows specifying a different corner radius for each corner of a rectangle separately. This means Top/Left can be different from Bottom/Right for example. However, it makes the assumption that all corner radius are circular instead of elliptical.
    - Originates from UWP's `Control`
    - This implementation/API was thought up new for UWP.
 3. `RadiusX`/`RadiusY` : This implementation only allows a uniform corner radius for all 4 corners of a rectangle. This means Top/Left is always going to match Bottom/Right and the other two corners. However, it does NOT assume a circular corner radius and allows an elliptical one instead.
    - Originates from WPF's `Rectangle` and `RectangleGeometry`
    - This implementation/API was likely carry-over from very old graphics APIs in Windows such as early DirectX or GDI.

As can be seen, each of these implementations is more powerful in a certain way than the other. However, each is missing some functionality the other one has.

**The discussion is: should CornerRadius be extended to support elliptical corner radius instead of assuming circular?** This would require 8 separate components in the struct as shown below. Doing this would significantly increase the power of corner radius and make its functionality complete in all cases. Note that if only TopLeft is specified, RadiusX/Y would be the same and it would function as circular. So backwards API compatibility is fully supported with only 4 components.

 1. TopLeft
    - TopLeftRadiusX
    - TopLeftRadiusY
 3. TopRight
    - TopRightRadiusX
    - TopRightRadiusY
 5. BottomRight
    - BottomRightRadiusX
    - BottomRightRadiusY
 6. BottomLeft
    - BottomLeftRadiusX
    - BottomLeftRadiusY

A CornerRadiusType property would also need to be added to clarify Circular/Elliptical. Perhaps IsCircular or IsElliptical could be used as well and calculated from X/Y values themselves. This would allow code to figure out the one special case: When code only supports 4 components (circular); however, radius X/Y were specified in the constructor (elliptical). In that case TopLeft, etc. would be averages of X/Y components and shouldn't really be used. This means:

 * Circular is backwards compatible with code that expects Elliptical
 * Elliptical is NOT always cleanly backwards compatible with code that expects Circular. In that case TopLeft, for example, would have to be averaged from TopLeftRadiusX/Y. So it is still fully functional but not always intuitive. Code should be able to detect if Elliptical radii are used.

There is a real opportunity here to make Avalonia more powerful and I would hate to pass it up. I'm hoping for feedback pretty quickly because if `CornerRadius` is updated I can update `Rectangle` as well pretty quickly. At this point I think it can all be done with basically no breaking changes!!

**Comment by timunie** [comment]
Back when I developed a WPF App I which I had this opportunity to have elliptical cornernradius at the top and Flat geometry at the bottom. I ended up with custom pathfigure.

So for me this would be a great addition 🙂

**Reply by robloo** [reply]
Yes, WPF never had a CornerRadius property but this does exist in UWP and now also Avalonia. It is possible using the existing property to specify a circular corner radius on the TopLeft and TopRight of a control while leaving the BottomLeft and BottomRight flat, right-angles (corner radius = 0).

**Comment by danielmayost** [comment]
@robloo You're doing an amazing job ... well done.
Even the core team is not keeping up.

**Reply by robloo** [reply]
Yea, don't assume my motives are purely altruistic though :) This is all related to ground work for a future commercial app and the impending 11.0 API lockdown is having to accelerate some things. Of course I always liked WPF conceptually though so I get into the weeds because this all interests me.

**Reply by maxkatz6** [reply]
To be honest I still think if you need highly customizable borders, you should go with Rectangle of Path instead. It is a relatively rare use case, not a big deal to replace default control template. 

But I won't stop anybody from implementing it, if somebody else want to approve it.

**Reply by robloo** [reply]
> To be honest I still think if you need highly customizable borders, you should go with Rectangle of Path instead.

Yes, I partly agree with you. However, a big part of the issue is with API design. We now have two different concepts for corner radius (to be fair so does UWP). These differences should be unified in my opinion. To unify them requires implemented a superset of the features. I think that can be done though without breaking things.

Edit: So what this spec allows is for CornerRadius to be used in Rectangle directly.

**Comment by robloo** [comment]
Added a proof-of-concept PR so you can take a look at the ideas in code: #7927

**Comment by robloo** [comment]
Note that an elliptical, non-uniform corner radii struct is already required to draw borders in Avalonia itself. We currently have two separate implementations:

 1. [`RoundedRect`](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/RoundedRect.cs)
 2. [`BorderGeometryKeypoints`](https://github.com/AvaloniaUI/Avalonia/blob/fe671b2cebf6f0a4b55e1a3a5a55cd7969c1122a/src/Avalonia.Controls/Utils/BorderRenderHelper.cs#L186) which is changed somewhat but still required with the BackgroundSizing PR

This again shows a gap in API-design and feature set that should have been addressed a long time ago.
